### PR TITLE
[WIP, RFC FS-1063] Add let! ... and! ... syntax

### PR DESCRIPTION
**Pull request by TD5**
Implementation of [F# RFC FS-1063 Applicatives syntax](https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1063-support-letbang-andbang-for-applicative-functors.md)

Not ready for prime time, just a proof-of-concept right now.

### CODE REVIEW

**Comment by TD5** [comment]
Suggestion: https://github.com/fsharp/fslang-suggestions/issues/579

File: src/fsharp/lex.fsl [hunk 1]
```diff
@@ -240,10 +240,14 @@ rule token args skip = parse
      { YIELD_BANG(false) } 
  | "match!"
      { MATCH_BANG }
+ | "and!"
+     { AND_BANG(false) }
+ | "anduse!"
```
Comment (dsyme) [inline src/fsharp/lex.fsl]: Interesting, I hadn't considered `anduse!`.

**Comment by dsyme** [comment]
This looks like very good work so far.  

* The `anduse!` surprised me, why not use `use!... and!`?   
* Are there cases where you want to mix `let!` and `use!` semantics in a single applicative?
* What is the semantics for `use! ... and!`?

**Comment by TD5** [comment]
> * The anduse! surprised me, why not use use!... and!?
> * Are there cases where you want to mix `let!` and `use!` semantics in a single applicative?

I don't know of any off the top of my head, but equally, I see no reason why it couldn't be the case that only a subset of arguments are disposable. It would be a bit odd and sad if someone found themselves having to fake a disposable to work around this, but if the community feels this is a case of YAGNI, then this can certainly change.

> * What is the semantics for `use! ... and!`?

This is mentioned in [the RFC pull request](https://github.com/fsharp/fslang-design/blob/cb99c9fbffaefeed5696f65a1ca3983bd7f05d82/RFCs/FS-1063-support-letbang-andbang-for-applicative-functors.md#managing-resources). The short answer is that it works just like `let! ... and! ...` but where the `let!` _may_ be swapped out for a `use!`, and any `and!` _may_ be swapped out for an `anduse!`. In any case where the swapping out occurs, we insert a call to `MapUsing : 'T * ('T -> 'U) -> 'U when 'U :> IDisposable` (the method name needs revisiting) logically inside the function of the applicative being applied (i.e. the lambda which is created when the `return` is desugared). In 99% of cases I expect `MapUsing` to just call the given function inside a `try ... finally`, disposing the `'T` afterwards.

**Comment by thinkbeforecoding** [comment]
When working on this for F#3, I also implemented a  for x in xs and y in ys do … better when the unerlying type as a collection meaning, like zipping arrays…
Would it be doable ?

**Comment by TD5** [comment]
@thinkbeforecoding, are you asking if I would implement something similar for query expressions too?

**Comment by thinkbeforecoding** [comment]
@TD5 Yes that was implemented for query expressions, on a fork:
https://github.com/thinkbeforecoding/fsharp

**Comment by TD5** [comment]
@thinkbeforecoding That is out of scope for the current change, as far as I am concerned. However, I see no reason why this could not be done, especially in light of the prior art. Have you considered creating an [fslang suggestion](https://github.com/fsharp/fslang-suggestions) for this if there isn't one already?

**Comment by cartermp** [comment]
@thinkbeforecoding I agree that this is out of scope for this PR (and corresponding RFC), but a further issue in fslang would be welcome. We can see if it makes sense to turn into an RFC and fold into the language separately from this PR.

**Comment by TD5** [comment]
Yep 👍 I’ve had my eye on it for a while and some theories as to what is going on but I’ve not been able to build the vs bits on my machine so it has been tricky.

**Comment by cartermp** [comment]
@TD5 If you install:

* .NET Desktop workload + F# desktop templates component
* Visual Studio Extensibility workload

And if you use the VS command line in admin mode with `build vs` does it succeed?

**Comment by TD5** [comment]
@cartermp, in the end I just installed everything under the sun and the tests are now running via `build.cmd`, so that's a big step forward.

**Comment by TD5** [comment]
Pull request tests don't seem to have kicked off - don't trust the supposed greenness of the build!

**Comment by TD5** [comment]
The tests run and fail locally as on CI when I run `build.cmd release ci_part1`, so that's working.

When I run in what I think is the `!FSCBINPATH!` directory (`visualfsharp\release\net40\bin`) and tack on `--test=Tests.LanguageService.AutoCompletion.UsingProjectSystem.CompletionInDifferentEnvs5` to the command that I think `build.cmd release ci_part1` kicks off:
```
"C:\Users\Tom\Documents\FSharpContribs\visualfsharp\packages\NUnit.Console.3.0.0\tools\nunit3-console.exe" --verbose --x86 --framework:V4.0 --result:"C:\Users\Tom\Documents\FSharpContribs\visualfsharp\tests\TestResults\test-vs-ideunit-results.xml;format=nunit3" --output:"C:\Users\Tom\Documents\FSharpContribs\visualfsharp\tests\TestResults\test-vs-ideunit-output.log"  --err:"C:\Users\Tom\Documents\FSharpContribs\visualfsharp\tests\TestResults\test-vs-ideunit-errors.log"  --work:"C:\Users\Tom\Documents\FSharpContribs\visualfsharp\release\net40\bin"  --workers=1 --agents=1 --full "C:\Users\Tom\Documents\FSharpContribs\visualfsharp\release\net40\bin\VisualFSharp.UnitTests.dll" --test=Tests.LanguageService.AutoCompletion.UsingProjectSystem.CompletionInDifferentEnvs5
```
I still see the same error as the one VS gives me:
```
NUnit Console Runner 3.0.5797
Copyright (C) 2015 Charlie Poole

Runtime Environment
   OS Version: Microsoft Windows NT 10.0.17134.0
  CLR Version: 4.0.30319.42000

Test Files
    C:\Users\Tom\Documents\FSharpContribs\visualfsharp\release\net40\bin\VisualFSharp.UnitTests.dll

Test Filters
    Test: Tests.LanguageService.AutoCompletion.UsingProjectSystem.CompletionInDifferentEnvs5

Agent process 4316 starting
Running under version 4.0.30319.42000, Net 4.5
Agent process 4316 exiting

Errors and Failures

1) TearDown Error : Tests.LanguageService.AutoCompletion.UsingProjectSystem
System.IO.FileNotFoundException : Could not load file or assembly 'Microsoft.VisualStudio.LanguageServices, Version=2.9.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
TearDown : System.NullReferenceException : Object reference not set to an instance of an object.
   at UnitTests.TestLib.LanguageService.LanguageServiceBaseTests.Init()
--TearDown
   at UnitTests.TestLib.LanguageService.LanguageServiceBaseTests.Cleanup()

2) Error : Tests.LanguageService.AutoCompletion.UsingProjectSystem.CompletionInDifferentEnvs5
OneTimeSetUp: System.IO.FileNotFoundException : Could not load file or assembly 'Microsoft.VisualStudio.LanguageServices, Version=2.9.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.

Run Settings
    RuntimeFramework: V4.0
    RunAsX86: True
    WorkDirectory: C:\Users\Tom\Documents\FSharpContribs\visualfsharp\release\net40\bin
    MaxAgents: 1
    NumberOfTestWorkers: 1
    Verbose: True

Test Run Summary
    Overall result: Failed
   Tests run: 1, Passed: 0, Errors: 1, Failures: 0, Inconclusive: 0
     Not run: 0, Invalid: 0, Ignored: 0, Explicit: 0, Skipped: 0
  Start time: 2018-10-23 10:49:14Z
    End time: 2018-10-23 10:49:14Z
    Duration: 0.215 seconds

Results (nunit3) saved as C:\Users\Tom\Documents\FSharpContribs\visualfsharp\tests\TestResults\test-vs-ideunit-results.xml
```

I'm sure I am just missing something basic like an environment variable since `build.cmd release ci_part1` works and it doesn't seem to clean up any build artefacts. Does anyone have any ideas?

**Comment by dsyme** [comment]
@TD5 - I updated this to master by push to your branch.  I'll aim to review it and use it next, and look at unresolved issues.

**Comment by TD5** [comment]
That’s excellent news! Thanks for taking an interest!

Let me know if you have any questions. From what I recall, the outstanding issue from my perspective was the effect on auto completion in VS (there are some tests labelled as something akin to “legacy Visual Studio” that are affected).

**Comment by dsyme** [comment]
I would like to revive this work and get it into F# 5.x

That's quite a lot of conflicts now, I'll see if I can resolve

**Comment by dsyme** [comment]
OK, merged with master

**Comment by dsyme** [comment]
I will resubmit this in a new PR

**Comment by dsyme** [comment]
Replaced by https://github.com/dotnet/fsharp/pull/7756 to make sure the feature comes from a feature branch

**Pull request closed without merging**